### PR TITLE
Add NewFx{{Service}}YARPCClient and NewFx{{Service}}YARPCProcedures generated methods from protoc-gen-yarpc-go for Fx

### PR DESCRIPTION
**Pull request by peter-edge**
This adds these two generated methods for every Protobuf service. We can later add a `--no-fx` flag for `protoc-gen-yarpc-go` if this becomes a problem. The `keyvaluefx.Client, keyvaluefx.Server` methods you are used to in Thrift are `foo.ProvideKeyValueYARPCClient, foo.ProvideKeyValueYARPCProcedures` in Protobuf. Generated Protobuf code by convention goes in the same Golang package as the source types code, unlike Thrift. 

Question: Do the `Params/Result` structs need to be public, or do just their fields need to be public? The structs are public in the corresponding Thrift code, but it would be nicer if they didn't have to be.

### CODE REVIEW

File: encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go [hunk 1]
```diff
@@ -30,12 +30,12 @@ import (
 	"go.uber.org/yarpc/internal/protoplugin"
 )
 
-const tmpl = `{{$packagePath := .GoPackage.Path}}
+const tmpl = `{{$packagePath := .GoPackage.Path}}{{$packageName := .GoPackage.Name}}
```
Comment (abhinav) [inline encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go]: FYI, if you're putting these on the same line just to have the "Code generated
by" at the top of the file, you can tell text/template to ignore leading/
trailing whitespace by using `{{-` and/or `-}}`.

```
const tmpl = `
{{- $packagePath := .GoPackage.Path -}}
{{- $packageName := .GoPackage.Name -}}
// Code generated by [..]
```

File: encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go [hunk 2]
```diff
@@ -193,6 +193,81 @@ func Build{{$service.GetName}}YARPCProcedures(server {{$service.GetName}}YARPCSe
 	)
 }
 
+// Provide{{$service.GetName}}YARPCClientParams defines the input
+// for Provide{{$service.GetName}}YARPCClient. It provides the
+// paramaters to get a {{$service.GetName}}YARPCClient in an
+// Fx application.
+type Provide{{$service.GetName}}YARPCClientParams struct {
```
Comment (abhinav) [inline encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go]: nit: Can we get a `$serviceName` variable here given that it's called 4 times in these 5 lines.

Comment (abhinav) [reply encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go]: Typically for Fx modules, we name the Params struct just `Params` or `FooParams`, not `ProvideFooParams`.

File: encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go [hunk 3]
```diff
@@ -193,6 +193,81 @@ func Build{{$service.GetName}}YARPCProcedures(server {{$service.GetName}}YARPCSe
...
+	fx.In
+
+	Provider yarpc.ClientConfig
+}
+
+// Provide{{$service.GetName}}YARPCClientResult defines the output
+// of Provide{{$service.GetName}}YARPCClient. It provides a
+// {{$service.GetName}}YARPCClient to an Fx application.
+type Provide{{$service.GetName}}YARPCClientResult struct {
```
Comment (abhinav) [inline encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go]: Same as above

File: encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go [hunk 4]
```diff
@@ -193,6 +193,81 @@ func Build{{$service.GetName}}YARPCProcedures(server {{$service.GetName}}YARPCSe
...
+}
+
+// Provide{{$service.GetName}}YARPCClient provides a {{$service.GetName}}YARPCClient
+// to an Fx application using the given name for routing.
+//
+//  fx.Provide(
+//    {{$packageName}}.Provide{{$service.GetName}}YARPCClient,
+//    ...
+//  )
+func Provide{{$service.GetName}}YARPCClient(name string, options ...protobuf.ClientOption) interface{} {
```
Comment (abhinav) [inline encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go]: If something is named `ProvideFoo`, the expectation is that it will return an
`fx.Option`, not a thing that you'll provide.

So this should either be renamed to just `func New{{serviceName}}YARPCClient`
or similar, or return an `fx.Option` by calling `fx.Provide`.

The former is preferable largely because it'll be confusing otherwise to do,

    fx.New(
      protopackage.ProvideFooYARPCClient("foo"),
      fx.Provide(
        thriftpackage.BarClient("bar")
      ),
    )

(You can't `fx.Provide` an `fx.Option`.)

Comment (peter-edge) [reply encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go]: We already have `New{{serviceName}}YARPCClient`, this is how you construct Clients without Fx. What alternative naming would you suggest?

File: encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go (hunk 2 above)
Comment (peter-edge) [reply encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go]: I can't call this `Params` because there can be multiple services per package. I can make this `FooYARPCClientParams`, but I figured it was more clear to name it `ProvideFooYARPCClientParams`. You're sure I should change this to the former? Preferably, I'd like it to match with whatever we name the function (which shouldn't be Provide as you said).

File: encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go (hunk 3 above)
Comment (peter-edge) [reply encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go]: Let's discuss in that comment

File: encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go (hunk 1 above)
Comment (peter-edge) [reply encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go]: Thanks for pro tip! If it's not a big deal, I'm just going to keep it as-is as this is how the rest of the template operates.

File: encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go (hunk 2 above)
Comment (peter-edge) [reply encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go]: Also, the `$service.GetName` - this is done in 100 other places in this template, I'd want it to be consistent - if it's a big deal, we can change this in a separate PR, but as you said I think it's a nit.

**Comment by peter-edge** [comment]
@abhinav replied:

1. What to name the functions instead of `Provide*`?
2. Can we have the struct naming match the function prefix, so if we s/Provide/Whatever, the structs are `Whatever*`?

On the structs being public:

```
Yeah, the params and results structs should be exported so that the dependencies and results are obvious when looking at the package docs.
```

I see it on the opposite side - since the majority of users don't use these structs, I don't think they should be public, and they just cause confusion. They've already confused someone reviewing this who isn't on the Fx team, case in point - since they're not used, it's not clear why they are public. Any such extensions can be easily done by just making your own Fx functions/structs.

I know that's probably not going to matter, and that we'll probably end up keeping them, but just my two cents.

File: encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go (hunk 1 above)
Comment (abhinav) [reply encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go]: Yeah, this was just an FYI.

File: encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go (hunk 2 above)
Comment (abhinav) [reply encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go]: Yeah, we can't call it `Params`. `FooYARPCClientParams` is preferable. They're never called `ProvideFooParams` because the constructors are never called `Provide*`.

Comment (peter-edge) [reply encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go]: See below comment though - we can't call the functions `New`, so I wanted to relate the structs and functions in some way, esp since the generated code returns `interface{}` from both functions.

File: encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go (hunk 4 above)
Comment (abhinav) [reply encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go]: That naming issue is why the Fx module gets its own subpackage. There's no
precedent here on whether Protobuf Fx modules get their own package or not. If
they get their own package, there's a lot more room around what it's called.

If it **has** to be in this package, there are two options:

- `ProvideFoo` returns an `fx.Option`
- `SomethingFooYARPCClient` where `Something` is TBD returns `interface{}`

What `Something` is, I don't know. I'm asking for some more deliberation
around Fx module structure for Proto before we commit to this.

Again, typically constructors are named `NewFoo` as in regular Go code but
that name is already taken -- which is why the Fx module was in a subpackage
in the first place.

File: encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go (hunk 2 above)
Comment (abhinav) [reply encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go]: I mean even if the constructor was called `NewFoo`, the params struct would still be called `FooParams`, not `NewFooParams`.

I think it's easy to say `FooParams` defines the parameters needed to build `Foo` regardless of whether the constructor for it is called `NewFoo` or `ProvideFoo`.

File: encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go (hunk 4 above)
Comment (peter-edge) [reply encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go]: Should we schedule a meeting to decide on what option to take?

This should be in the same package, this is the standard Protobuf plugin convention. I'd like to decide this however, if a meeting is the quickest path, then let's set that up.

**Comment by abhinav** [comment]
> 1. What to name the functions instead of Provide*?

I don't know. I'm asking for more deliberation and more team input to be
solicited before we commit to this. My preference would be for the Fx module
to be its own subpackage rather than putting everything in the same package.
I'm not a Protobuf expert so I'll defer to you and others on that discussion,
but I don't see how the approach of putting everything in the same package is
scalable. What if we want to generate mock clients?

> 2. Can we have the struct naming match the function prefix, so if we s/Provide/Whatever, the structs are Whatever*?

I don't think we should do that. Parameter objects are a pretty standard
design pattern which we've slightly adapted for Fx. Typically, a parameter
object is named after the function/method it is for with an extra suffix/
prefix (which happens to be `Params` and `Result` in Fx).

When a parameter object is for a constructor, it is named after the type it's
building. In Go, even though we call the function `NewFoo`, I think we can
agree that it's a constructor for the type `Foo`, so the parameter object for
it should be `FooParam`.

We're basically generating an Fx module here, so we shouldn't go too far off
the path of normal Fx modules.

---

RE: Public/private structs. The reason normal Fx modules export these structs
is that being able to call the constructor directly to build the underlying
objects has always been a part of the design. With Thrift/Proto clients, we
have some leeway because we hide the constructors. That said, I think it's
preferable for this and Thrift to err on the side of looking like a normal Fx
module where possible than "here are a couple of Fx helper functions".

**Comment by peter-edge** [comment]
Talked offline, going to look to talk about this next week when I'm in town.

Tldr it's really against normal Protobuf design conventions to put this in a separate package, but I'm open to that we may have to make an exception here.

**Comment by kriskowal** [comment]
My sentiments are similar to @abhinav.

If the fx package is separate, these would be consistent with Thrift:

- {{service}}pbfx/Client
- {{service}}pbfx/Server

To summarize what you’ve both suggested, If we took a step away from that and pivoted the service name into the package, we get the name conflict.

- {{service}}pb/{{Service}}Client
- {{service}}pb/{{Service}}Server

There is no precedent for Provide in the name. fx.Provide takes a constructor and returns a module or option. This is a function that takes nothing (or the corresponding constructor by implication) and returns a module or option.

There is a precedent to export a single option called `Module`. We could extend that precedent by returning client and server modules with the service prefix.

- {{service}}pb/{{Service}}ClientModule
- {{service}}pb/{{Service}}ServerModule

**Comment by peter-edge** [comment]
@kriskowal @breerly I updated the functions to `NewFx` per our conversation yesterday. @abhinav tldr due to some wrinkles with protoc and Golang protoc plugins out in the wild, separating this into a separate package is an extremely difficult and potentially impossible (in a backwards-compatible way) task. I can explain more next week.
